cpp/src/progress_thread.cpp

```diff
@@ -102,7 +102,11 @@ void ProgressThread::remove_function(FunctionID function_id) {
     functions_.erase(function_id.function_index);
 
     if (functions_.empty()) {
-        thread_.pause();
+        // Pausing the thread is not atomic, and so `event_loop` may
+        // run, needing the lock, we must therefore drop the lock before
+        // pausing.
+        lock.unlock();
+        thread_.pause_nb();
     }
```

Comment:
I think this comment is true if we use the blocking `thread_.pause()` (which definitely waits for the iteration to complete with us holding the lock). With nonblocking pause I don't think we need to drop the lock. (This was what was saving us before).

Reply:
I put the comment in when I was using `pause` rather than `pause_nb`...